### DISCUSSION TITLE
✨ Support preemptible execution environments

### DISCUSSION THREAD

**Discussion by davidwallacejackson** [dagger/dagger#12524] [Ideas]
### What are you trying to do?

I've recently been considering moving pieces of my existing (non-Dagger) CI pipeline to self-hosted infrastructure on Google's spot instances. Because these instances could be preempted at any time, it would be ideal if progress on my pipeline were saved incrementally, in as granular a way as possible. It occurred to me that Dagger's cache model (where each execution step can be cached, potentially to long-lived external storage) might make solving this problem a lot easier.

I think the ideal solution is to be able to export cache incrementally, throughout the execution of a larger pipeline and in parallel with other pipeline steps.

### Why is this important to you?

Spot instances are 50-90% cheaper than reserved compute. Dagger already provides incredible opportunities to reduce spend on CI through caching, but making it easy to use preemptible environments could offer massive savings on the work that actually does have to get done.

**Comment by shykes** [comment]
That's very interesting. I wonder how much you could get out of the current version of Dagger on spot instances, with either a regular cache export at the end of the job, or persistent block storage to avoid the need for cache export entirely? Of course some jobs would be interrupted before the end, but depending on the average lifespan of a spot instance, it might happen infrequently enough to be worth it.

Regardless of caching, there are other aspects of this setup that would not be trivial: load-balancing / proxying; and retries on failure. Have you given those any thought?

As for the caching improvement you are describing: it's possible but will probably require non-trivial engineering both on dagger and upstream buildkit as well. We have an ambitious roadmap for dagger's caching infrastructure though, so we'll get around to it eventually - but not in the near future I believe.

**Comment by vesperrin** [comment]
Hi,

Are you able to give more details about this?
>  We have an ambitious roadmap for dagger's caching infrastructure though, so we'll get around to it eventually - but not in the near future I believe.
